src/os/bluestore/BlueStore.cc

```diff
@@ -3560,11 +3560,16 @@ void BlueStore::MempoolThread::_trim_shards(bool interval_stats)
 void BlueStore::MempoolThread::_tune_cache_size(bool interval_stats)
 {
   auto cct = store->cct;
-  uint64_t target = store->osd_memory_target;
   uint64_t base = store->osd_memory_base;
-  double fragmentation = store->osd_memory_expected_fragmentation;
-  uint64_t cache_max = ((1.0 - fragmentation) * target) - base;
```

Comment:
FWIW, we should at least confirm that:
- cache_max won't be overflowed
- cache_max >= cache_min

Reply:
We should absolutely fix the overflow.  Beyond that, are there other issues that are causing problems for you?  I think we should absolutely try to adhere to the user request where possible.  That's why I'm not convinced we want to muck with the target.

Reply:
@markhpc

> Edit: Ok, I see where you are going with this. I think I had misinterpreted the point of it earlier. What would you think about something like this? (This reads easier for me, but perhaps not others?)
> 
> uint64_t cache_max = cache_min;
> uint64_t limited_target = (1.0 - fragmentation) * target;
> if (limited_target > base + cache_min) {
> cache_max = limited_target - base;
> }

It's not enough.
I think a more complete fix would be something like:
- if the __osd_memory_target__ is big enough, then we are all good.
- if we don't have enough memory in hand, e.g,  __osd_memory_target__ < (__osd_memory_cache_min__ + __osd_memory_base__), then:
  - we did not cache anything for bluestore at all, or:
  - we still specify a base cut-off ratio for the bluestore cache, e.g., 0.2 *  __osd_memory_target__ or at least 64MB perhaps? (Of course the numbers here are still TBD)

Reply:
> @markhpc
> 
> > Edit: Ok, I see where you are going with this. I think I had misinterpreted the point of it earlier. What would you think about something like this? (This reads easier for me, but perhaps not others?)
> > uint64_t cache_max = cache_min;
> > uint64_t limited_target = (1.0 - fragmentation) * target;
> > if (limited_target > base + cache_min) {
> > cache_max = limited_target - base;
> > }
> 
> It's not enough. If we don't want any kind hardcaps of **osd_memory_target**, then
> I think a more complete fix would be something like:
> 
>     * if the **osd_memory_target** is big enough, then we are all good.
> 
>     * if we don't have enough memory in hand, e.g,  **osd_memory_target** < (**osd_memory_cache_min** + **osd_memory_base**), then:
>       
>       * we simply do not cache anything for bluestore at all, or:
>       * we still specify a basic cut-off ratio for the bluestore cache, e.g., 0.2 *  **osd_memory_target** or at least 64MB perhaps? (Of course the numbers here are still TBD)

if (limited) osd_memory_target < (osd_memory_cache_min + osd_memory_base), I don't think we should do anything other than observe the cache_min.  We don't want to set it higher because base is our estimate of what the OSD would use once in a steady state, and we don't want to set it lower because cache_min is the smallest aggregate cache size bluestore is told it can work with.  Am I missing something that breaks beyond the uint64_t subtraction issue?

Reply:
> uint64_t limited_target = (1.0 - fragmentation) * target;

Shall we at least restrict fragmentation to [0, 1.0) ?

Reply:
@markhpc Repushed. Can you recheck?